File: QuickSort.java
public class QuickSort{

    private static void swap(int[] array, int i, int j){
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    public static void quickSort(int[] array, int left, int right){
        if (left < right){
            int pivot = array[right];
            int i = left - 1;

            for (int j = left; j < right; j++){
                if (array[j] <= pivot){
                    i++;
                    swap(array, i, j);
                }
            }
            swap(array, i + 1, right);

            int partitionIndex = i + 1;

            // Recursividade
            quickSort(array, left, partitionIndex - 1);
            quickSort(array, partitionIndex + 1, right);
        }
    }

    public static void printArray(int[] array){
        for (int num : array){
            System.out.print(num + " ");
        }
        System.out.println();
    }

    public static void main(String[] args){
        int[] array = {10, 7, 8, 9, 1, 5};
        int n = array.length;

        System.out.println("Array original:");
        printArray(array);

        quickSort(array, 0, n - 1);

        System.out.println("Array ordenado:");
        printArray(array);
    }
}

File: quickSort.cpp
#include <iostream>
using namespace std;

void swap(int& a, int& b){
    int temp = a;
    a = b;
    b = temp;
}

// Função para particionar o array
int partition(int arr[], int left, int right){
    int pivot = arr[right]; // Escolhe o pivô
    int i = (left - 1); // Índice do menor elemento

    for (int j = left; j <= right - 1; j++){
        // Se o elemento atual for menor ou igual ao pivô
        if (arr[j] <= pivot){
            i++; // Incrementa o índice do menor elemento
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[right]);
    return (i + 1);
}

// Função principal que implementa o QuickSort
void quickSort(int arr[], int left, int right){
    if (left < right) {
        // pi é o índice de particionamento, arr[pi] está no lugar certo
        int pi = partition(arr, left, right);

        // Separa recursivamente os elementos antes e depois da partição
        quickSort(arr, left, pi - 1);
        quickSort(arr, pi + 1, right);
    }
}

void printArray(int arr[], int size){
    for (int i = 0; i < size; i++){
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main(){
    int arr[] = {10, 7, 8, 9, 1, 5};
    int n = 6;
    quickSort(arr, 0, n - 1);
    cout << "Array ordenado: ";
    printArray(arr, n);
    return 0;
}
